\documentclass[12pt]{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsfonts}
\usepackage{fancyvrb}
\usepackage{lstlisting}

\usepackage[margin=1in]{geometry}


\begin{document}

\section{The BED Format Specification}

\subsection{An example}

\subsection{Terminologies and Concepts}

\subsubsection{Character set restrictions}

ASCII? Unicode?

\subsection{Header lines}

Header lines are used to define a custom track for use in the Genome Browser.

\subsubsection{Browser lines}

Browser lines configure the overall display of the Genome Browser window when your annotation file is uploaded. Each line defines one display attribute. (Copied from https://genome.ucsc.edu/goldenPath/help/customTrack.html)

\begin{tabular}{ |c|c| }
\hline
Attribute Name & Description \\
\hline
position <position> & (Description) Accepted format: \lstinline{chromosone:start-end}
\begin{Verbatim}[frame=single]
  (Some regex) e.g. ^chromosone:\d+\-\d+\$
\end{Verbatim}
and \lstinline{start} \leq \lstinline{end}.
\hline
hide all & (text)
\hline
\end{tabular}

\subsubsection{Track lines}

\begin{tabular}{ |c|c| }
\hline
Attribute & Description \\
\hline
name=<track_label> & (description) Some alpha-numeric string [a-z0-9A-Z]{1, 15} \\
description=<center_label> & (description) Some alpha-numeric string [a-z0-9A-Z]{1,60}
\hline
\end{tabular}

\subsection{BED fields}

% Idea for chrom: ((C|c)hr((\d+)|X|Y|M)(\_random)?)|(scaffold\_\d+)
% Idea for chromStart: [0, N] where N is the length of the longest chromosome
% Idea for chromEnd: [1, N+1]
% Idea for name: .*
% Idea for score: [0, 1000]
% Idea for strand +|-|.
% Idea for thickStart: [chromStart, chromEnd]
% Idea for thickEnd: [chromStart, chromEnd] but thickStart <= thickEnd
% Idea for itemRgb: [0, 255],[0, 255],[0, 255] 
%      In regex: (?(0|1) \d{1,2} | (?(2) (?([0-4]) \d? | (?(5) [0-5]? | \d?)) 
%                  | \d?)), {1,3}
% Idea for blockCount: [0, ?]
% Idea for blockSizes: (\d+,)*{blockCount-1}\d+
% Idea for blockStarts: (\d+,)*{blockCount-1}\d+

Fields 1 to 3 are mandatory, fields 4 to 12 are optional but if a field is
used, all other fields before it must also be filled.

\begin{tabular}{c c c c c}
Col & Field & Type & Regexp/Range & Brief Description \\
\hline
1 & chrom & String & (Some regex) (Chr|chr)(01-no idea|X|Y|M) or others & (Some Description) \\
2 & chromStart & Int & [0, N) & (Some Description) \\
3 & chromEnd & Int & [1, N) & (Some Description) \\
4 & name & String & (Some regex) & (Some Description)
5 & score & Int & [0, 1000] & (Some Description) \\
6 & strand & String & +|-|. & (Some Description)
\hline
\end{tabular}

% These are the more detailed descriptions of each of the attributes
\begin{enumerate}
  \item \textbf{chrom}: The name of the chromosome or scaffold.
    % Specify that it's  mandatory, and that it's format must match the format 
    % of a chromosome from a reference genome, such has hg19
  \item \textbf{chromStart}: The starting position of the feature in the 
    chromosome or scaffold.
    % Specify that the value should be within the range of the chromosome that
    % is present in the chrom attribute.
  \item \textbf{chromEnd}: The ending position of the feature in the chromosome
    or scaffold.
    % Specify that it must be strictly greater than chromStart. 
    % Also recall that the chromosome indices are half-open, meaning that the
    % base at index <chromEnd> won't be included in any visualization.
  \item \textbf{name}: Defines the name of the BED line. (Optional)
    % Mention that the name is displayed in Genome Browser
    % Does the name have to be one word? (I imagine parsers won't be able to
    % handle quotation marks
  \item \textbf{score}: A score between 0 and 1000. (Optional)
    % This score is generated from any tool
  \item \textbf{strand}: Defines the strand.
    % Just whether the feature is on the coding or anticoding strand
  \item \textbf{thickStart}: The starting position at which the feature
    is drawn thickly.
    % Specify that this is just for visualization in the Genome Browser
    % Specify that if there is no thick part, then make thickStart equal
    % to thickEnd
  \item \textbf{thickEnd}: The ending position at which the feature is drawn
    thickly.
    % Specify that thickEnd >= thickStart
  \item \textbf{itemRgb}: A RGB value that determines the color that will
    be displayed with this feature.
    % Specify the format of R,G,B
    % Specify that for this to be useful, need itemRgb='On' in the track line.
  \item \textbf{blockCount}: The number of exons in the BED line.
  \item \textbf{blockSizes}: A comma-separated list of the block sizes.
    % Specify list should have length blockCount
    % Specify that the blockSize should not be larger than size of the
    % feature.
  \item \textbf{blockStarts}: A comma-separated list of block starts.
    % Specify that the starts are calculated relative to the chromStart value.
    % i.e. 0 means the block starts at chromStart.
    % Specify that the start value should be between 0 and
    % chromEnd - blockSize.
\end{enumerate}
\section{Recommended Practice for the BED Format}

\end{document}
